[JSC] Optimize `String.prototype.at` like `String.prototype.charAt`

sosukesuzuki opened the pull request:
#### 95e9295e5ff3bb36eb40c8a1e4053bce85e00f03
<pre>
[JSC] Optimize `String.prototype.at` like `String.prototype.charAt`
<a href="https://bugs.webkit.org/show_bug.cgi?id=217139">https://bugs.webkit.org/show_bug.cgi?id=217139</a>

Reviewed by NOBODY (OOPS!).

This patch optimizes `String.prototype.at`. It reuses most of the optimizations made for `String.prototype.charAt`, although there are some differences according to the specifications[1][2]:

- Returns undefined instead of an empty string when out of bounds
- Treats the `index` as `length + index` when the `index` is negative

                                            TipOfTree                  Patched

string-prototype-at-negative-index        2.3190+-0.0785     ^      1.6215+-0.1655        ^ definitely 1.4302x faster
string-prototype-at-out-of-bounds         1.7695+-0.1920            1.6320+-0.0462          might be 1.0843x faster
string-prototype-at-positive-index        1.9503+-0.1246     ^      1.6083+-0.1145        ^ definitely 1.2126x faster

[1]: <a href="https://tc39.es/ecma262/#sec-string.prototype.charat">https://tc39.es/ecma262/#sec-string.prototype.charat</a>
[2]: <a href="https://tc39.es/ecma262/#sec-string.prototype.at">https://tc39.es/ecma262/#sec-string.prototype.at</a>

* JSTests/microbenchmarks/string-prototype-at-negative-index.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-out-of-bounds.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-positive-index.js: Added.
(test):
* JSTests/stress/string-prototype-at.js: Added.
(shouldBe):
(testZeroIndex):
(testPositiveIndex):
(testOutOfBounds):
(testNegativeIndex):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(at): Deleted.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByValOnString):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileStringCharAtImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileStringAt):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95e9295e5ff3bb36eb40c8a1e4053bce85e00f03

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50650 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44022 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33051 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24662 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39025 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48551 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41507 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20326 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22319 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42684 "Found 1 new test failure: fast/css/viewport-height-border.html (failure)") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6016 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41262 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44330 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43117 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52544 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47461 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23012 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19360 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46334 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24281 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45387 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25075 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54958 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24003 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11302 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -1083,6 +1084,8 @@ class FixupPhase : public Phase {
             blessArrayOperation(node->child1(), node->child2(), node->child1()); // Rewrite child1 with ResolveRope.
             fixEdge<KnownStringUse>(node->child1());
             fixEdge<Int32Use>(node->child2());
+            if (node->op() == StringAt && m_graph.hasExitSite(node->origin.semantic, OutOfBounds))
```
Fixup phase is after prediction propagation. So this is late. You need to check this status and update array mode in DFGByteCodeParser.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -3203,6 +3203,16 @@ void SpeculativeJIT::compileGetByValOnString(Node* node, const ScopedLambda<std:
+
     // unsigned comparison so we can filter out negative indices and indices that are too large
     loadPtr(Address(baseReg, JSString::offsetOfValue()), scratchReg);
     Jump outOfBounds = branch32(
```
I think you should add outOfBounds + StringAt + OutOfBounds handling, which just puts undefined to the result.

sosukesuzuki replied:
The following code doesn't work (test for outOfBounds fails). Am I doing something wrong?

```cpp
    if (node->op() == StringAt && node->arrayMode().isOutOfBounds()) {
        outOfBounds.link(this);
        moveValue(jsUndefined(), resultRegs);
        doneCases.append(jump());
    }
```

Constellation commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -9975,16 +9978,19 @@ IGNORE_CLANG_WARNINGS_END
-            m_out.aboveOrEqual(
-                index, m_out.load32NonNegative(stringImpl, m_heaps.StringImpl_length)),
+            m_out.aboveOrEqual(index, stringLength),
             rarely(slowPath), usually(fastPath));
```
Ditto. slowPath should just fill undefined.

Constellation commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -970,6 +971,39 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncToString, (JSGlobalObject* globalObject,
+    JSValue argument0 = callFrame->argument(0);
+    if (argument0.isInt32()) {
+        int32_t i = argument0.asInt32();
+        double k = i < 0 ? length + i : i;
```
Let's just use `int64_t`. Avoid using double here.
```
int64_t k = i < 0 ? static_cast<int64_t>(length) + i : i;
```

Commit:
[JSC] Optimize `String.prototype.at` like `String.prototype.charAt`
https://bugs.webkit.org/show_bug.cgi?id=217139

Reviewed by NOBODY (OOPS!).

This patch optimizes `String.prototype.at`. It reuses most of the optimizations made for `String.prototype.charAt`, although there are some differences according to the specifications[1][2]:

- Returns undefined instead of an empty string when out of bounds
- Treats the `index` as `length + index` when the `index` is negative

                                            TipOfTree                  Patched

string-prototype-at-negative-index        2.3190+-0.0785     ^      1.6215+-0.1655        ^ definitely 1.4302x faster
string-prototype-at-out-of-bounds         1.7695+-0.1920            1.6320+-0.0462          might be 1.0843x faster
string-prototype-at-positive-index        1.9503+-0.1246     ^      1.6083+-0.1145        ^ definitely 1.2126x faster

[1]: https://tc39.es/ecma262/#sec-string.prototype.charat
[2]: https://tc39.es/ecma262/#sec-string.prototype.at

* JSTests/microbenchmarks/string-prototype-at-negative-index.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-out-of-bounds.js: Added.
(test):
* JSTests/microbenchmarks/string-prototype-at-positive-index.js: Added.
(test):
* JSTests/stress/string-prototype-at.js: Added.
(shouldBe):
(testZeroIndex):
(testPositiveIndex):
(testOutOfBounds):
(testNegativeIndex):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(at): Deleted.
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
* Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp:
(JSC::DFG::BackwardsPropagationPhase::propagate):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasArrayMode):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByValOnString):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileStringCharAtImpl):
(JSC::FTL::DFG::LowerDFGToB3::compileStringAt):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
